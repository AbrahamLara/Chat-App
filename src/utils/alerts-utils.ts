export enum ErrorPayloadType {
  REGISTER = 'register',
  LOGIN = 'login',
}

/**
 * An alert message object.
 */
export interface AlertMessage {
  /**
   * The message to display.
   */
  message: string;
}

/**
 * Describes an alert banner that displays messages for the user to see on a page.
 */
export interface AlertBarOptions extends AlertMessage {
  /**
   * Determines if the banner should be visible.
   */
  visible: boolean;

  /**
   * The type of alert.
   */
  type: string;
}

/**
 * Describes an error message for an invalid form field value.
 */
export interface FormAlertMessage extends AlertMessage {
  /**
   * The form field the error message is for. This may not to be provided if a form error wasn't due to an invalid field
   * value.
   */
  field?: string;
}

/**
 * Describes the payload to return for a failed request.
 */
export interface ErrorPayload {
  /**
   * The error type.
   *
   * Note: This might not actually be needed.
   */
  type: ErrorPayloadType;

  /**
   * The error messages produced. Right now the backend only serves form errors.
   */
  errors: FormAlertMessage[];
}

export const DEFAULT_ALERT_BAR_OPTIONS: AlertBarOptions = {
  visible: false,
  type: '',
  message: '',
};

/**
 * Determines if the give value is an error response object sent from a server request.
 */
export function isFormErrorMessage(response: any) {
  return typeof response === 'object' && response.type && response.errors;
}

/**
 * Creates a generic message object to display as an alert.
 */
export function createAlertMessage(message: string): AlertMessage {
  return { message };
}
