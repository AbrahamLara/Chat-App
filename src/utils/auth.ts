export interface RegexRule {
  regex: RegExp;

  rule: string;

  passed?: boolean;
}

/**
 * Password regex also used in the backend.
 * [PASSWORD_REGEX](https://github.com/AbrahamLara/chat-app-backend/blob/6757dd186701f4581f931c957d8bccd7840d4d3d/utils/misc.ts#L22-L41)
 */
export const PASSWORD_REGEX = /^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[!@#$%^&*\-_]).{8,}$/;

/**
 * Email regex also used in the backend.
 * [EMAIL_REGEX](https://github.com/AbrahamLara/chat-app-backend/blob/6757dd186701f4581f931c957d8bccd7840d4d3d/utils/misc.ts#L43-L48)
 */
const EMAIL_REGEX = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;

export const EMAIL_FORM_RULES = [
  (email: string) => !!email || 'Email is required',
  (email: string) => EMAIL_REGEX.test(email) || 'Email must be valid.',
];

export const PASSWORD_FORM_RULES = [
  (password: string) => !!password || 'Password is required',
  (password: string) =>
    PASSWORD_REGEX.test(password) || 'Password must be valid.',
];

export const PASSWORD_REGEX_RULES: RegexRule[] = [
  { rule: '8 characters minimum', regex: /.{8,}/ },
  { rule: '1 uppercase character', regex: /[A-Z]/ },
  { rule: '1 lowercase character', regex: /[a-z]/ },
  { rule: '1 number', regex: /[0-9]/ },
  {
    rule: '1 special character: !, @, #, $, %, ^, &, *, -, _',
    regex: /[!@#$%^&*\-_]/,
  },
];
