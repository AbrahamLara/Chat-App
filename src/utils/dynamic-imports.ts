// This file contains dynamic imports of components to reduce bundle size when building project with webpack.

const getAbout = () =>
  import(/* webpackChunkName: "About" */ '@/components/About.vue');

const getAppBar = () =>
  import(/* webpackChunkName: "AppBar" */ '@/components/AppBar.vue');

const getAuthPage = () =>
  import(/* webpackChunkName: "AuthPage"  */ '@/views/AuthPage.vue');

const getFormCard = () =>
  import(/* webpackChunkName: "FormCard" */ '@/components/FormCard.vue');

const getNotFound = () =>
  import(/* webpackChunkName: "NotFound" */ '@/views/NotFound.vue');

const getRegexRulesTester = () =>
  import(
    /* webpackChunkName: "RegexRulesTester" */ '@/components/RegexRulesTester.vue'
  );

const getSettings = () =>
  import(/* webpackChunkName: "Settings" */ '@/components/Settings.vue');

const getSignIn = () =>
  import(/* webpackChunkName: "SignIn" */ '@/components/SignIn.vue');

const getSignUp = () =>
  import(/* webpackChunkName: "SignUp" */ '@/components/SignUp.vue');

const getSplash = () =>
  import(/* webpackChunkName: "Splash" */ '@/views/Splash.vue');

export {
  getAbout,
  getAppBar,
  getAuthPage,
  getFormCard,
  getNotFound,
  getRegexRulesTester,
  getSettings,
  getSignIn,
  getSignUp,
  getSplash,
};
