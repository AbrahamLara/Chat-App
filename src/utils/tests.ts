/**
 * Resolves a vuetify warning of not being able to locate a data-app attribute on an element because vuetify
 * components should be wrapped by a v-app component.
 *
 * Ex: [Vuetify] Unable to locate target [data-app] in "v-menu"
 */
function resolveVuetifyAppDataWarning() {
  const app = document.createElement('div');
  app.setAttribute('data-app', 'true');
  document.body.append(app);
}

export { resolveVuetifyAppDataWarning };
