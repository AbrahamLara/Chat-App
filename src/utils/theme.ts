// Themes as string enums since the app state theme should only be either of these string values.
export enum THEME {
  DARK = 'Dark',
  LIGHT = 'Light'
}

// The application theme background color hexes.
export const ThemeHexes = {
  [THEME.DARK]: '#121212',
  [THEME.LIGHT]: '#fff'
};
