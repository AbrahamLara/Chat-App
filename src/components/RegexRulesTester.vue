<template>
  <v-sheet>
    <div
      :key="result.rule"
      v-for="result in results"
      :class="result.passed ? 'success--text' : 'error--text'"
    >
      <v-icon :color="result.passed ? 'success' : 'error'">{{
        result.passed ? 'mdi-check-circle' : 'mdi-alert'
      }}</v-icon>
      {{ result.rule }}
    </div>
  </v-sheet>
</template>

<script lang="ts">
import Vue from 'vue';
import { Component, Prop } from 'vue-property-decorator';
import { RegexRule } from '@/utils/auth-utils';

@Component({ name: 'RegexRulesTester' })
export default class RegexRulesTester extends Vue {
  @Prop({ type: Array }) rules!: RegexRule[];
  @Prop({ type: String, default: '' }) text!: string;

  /**
   * The results of running each test.
   */
  get results() {
    return this.rules.map(rule => ({
      ...rule,
      passed: RegExp(rule.regex).test(this.text),
    }));
  }
}
</script>
