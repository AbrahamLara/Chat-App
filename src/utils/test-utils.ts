import VueRouter, { RouteConfig } from 'vue-router';
import Vuex from 'vuex';
import { AppState } from '@/store/store-states';
import appMutations from '@/store/mutations';
import appActions from '@/store/actions';
import appModules from '@/store/modules';
import { DEFAULT_ROOT_STATE } from '@/store';
import { FormAlertMessage } from '@/utils/alerts-utils';

/**
 * Resolves a vuetify warning of not being able to locate a data-app attribute on an element because vuetify
 * components should be wrapped by a v-app component.
 *
 * ```
 * Ex: [Vuetify] Unable to locate target [data-app] in "v-menu"
 * ```
 */
export function resolveVuetifyAppDataWarning() {
  const app = document.createElement('div');
  app.setAttribute('data-app', 'true');
  document.body.append(app);
}

/**
 * Create a Vuex store instance for testing.
 *
 * TODO: Reuse to create store in main.ts.
 *
 * @param vx The Vuex value
 * @param customState A state object that modifies the default state
 */
export function createStore(
  vx: typeof Vuex,
  customState: Partial<AppState> = {}
) {
  const mutations = appMutations;
  const actions = appActions;
  const modules = appModules;

  return new vx.Store({
    state: { ...DEFAULT_ROOT_STATE, ...customState },
    actions,
    mutations,
    modules,
  });
}

/**
 * Creates a VueRouter instance for testing.
 *
 * TODO: Reuse to create router in main.ts.
 *
 * @param vr The VueRouter value
 * @param routes The route config. If not provided, then a default config is provided.
 */
export function createRouter(
  vr: typeof VueRouter,
  routes: RouteConfig[] = [{ path: '/' }]
) {
  return new vr({ routes, mode: 'history' });
}

/**
 * Creates a form error message for testing.
 */
export function createFormAlertMessage(
  field: string,
  message: string
): FormAlertMessage {
  return { field, message };
}
