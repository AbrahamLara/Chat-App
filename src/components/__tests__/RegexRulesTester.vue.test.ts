import { createLocalVue, mount, Wrapper } from '@vue/test-utils';
import Vue from 'vue';
import Vuetify from 'vuetify';
import RegexRulesTester from '@/components/RegexRulesTester.vue';
import { PASSWORD_REGEX_RULES } from '@/utils/auth-utils';

const localVue = createLocalVue();
Vue.use(Vuetify);

describe('RegexRulesTester.vue', () => {
  let wrapper: Wrapper<any>;

  beforeAll(() => {
    const vuetify = new Vuetify();

    wrapper = mount(RegexRulesTester, {
      localVue,
      vuetify,
      propsData: { rules: PASSWORD_REGEX_RULES, text: 'Testing1!' },
    });
  });

  it('handles valid password', async () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
