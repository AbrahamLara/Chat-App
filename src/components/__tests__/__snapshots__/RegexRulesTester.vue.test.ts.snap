// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegexRulesTester.vue handles valid password 1`] = `
<div class="v-sheet theme--light">
  <div class="success--text"><i aria-hidden="true" class="v-icon notranslate mdi mdi-check-circle theme--light success--text"></i>
    8 characters minimum
  </div>
  <div class="success--text"><i aria-hidden="true" class="v-icon notranslate mdi mdi-check-circle theme--light success--text"></i>
    1 uppercase character
  </div>
  <div class="success--text"><i aria-hidden="true" class="v-icon notranslate mdi mdi-check-circle theme--light success--text"></i>
    1 lowercase character
  </div>
  <div class="success--text"><i aria-hidden="true" class="v-icon notranslate mdi mdi-check-circle theme--light success--text"></i>
    1 number
  </div>
  <div class="success--text"><i aria-hidden="true" class="v-icon notranslate mdi mdi-check-circle theme--light success--text"></i>
    1 special character: !, @, #, $, %, ^, &amp;, *, -, _
  </div>
</div>
`;
